src/ast/passes/semantic_analyser.cpp

```diff
@@ -720,10 +724,49 @@ void SemanticAnalyser::visit(Call &call)
   } else if (call.func == "delete") {
     check_assignment(call, false, false, false);
     if (check_varargs(call, 1, std::numeric_limits<size_t>::max())) {
-      for (const auto *arg : call.vargs) {
-        if (!arg->is_map)
-          LOG(ERROR, arg->loc, err_)
-              << "delete() only expects maps to be provided";
+      if (call.vargs.size() == 2) {
+        auto *key_arg = call.vargs.at(1);
+        if (key_arg->is_map) {
+          // This supports the deprecated delete API of passing multiple maps as
+          // args
+          if (!call.vargs.at(0)->is_map)
+            LOG(ERROR, call.vargs.at(0)->loc, err_) << DELETE_ERROR;
+        } else {
+          if (!call.vargs.at(0)->is_map)
+            LOG(ERROR, call.vargs.at(0)->loc, err_) << DELETE_ERROR;
+
+          Map &map = static_cast<Map &>(*call.vargs.at(0));
+          if (map.key_expr) {
+            LOG(ERROR, call.vargs.at(0)->loc, err_)
+                << "delete() expects a map with no keys for the first argument";
+          } else {
+            auto search_key = map_key_.find(map.ident);
+            if (search_key != map_key_.end()) {
+              auto key_type = create_key_type(call.vargs.at(1)->type,
+                                              call.vargs.at(1)->loc);
+              if (!search_key->second.IsSameType(key_type) ||
+                  !key_type.FitsInto(search_key->second)) {
+                LOG(ERROR, call.vargs.at(1)->loc, err_)
+                    << "Argument mismatch for " << map.ident << ": "
+                    << "trying to delete with key of type: '"
+                    << call.vargs.at(1)->type << "' when map has key of type: '"
+                    << search_key->second << "'";
+              }
+            }
+          }
+
+          map.key_expr = call.vargs.at(1);
+          call.vargs.pop_back();
```

Comment:
So after the AST is manipulated here the rest of the 9 passes leading up to `is_final_pass()` fall into the deprecated multi-map delete codepath below. That's a little tricky but not wrong I suppose. Mind updating the comment below so that's captured?

Reply:
Yeah, that's a good point. I think this is the easiest way to maintain backward compat but I'll add a comment.